src/device-manager/java/src/nl/Weave/DataManagement/WDMClientImpl.java

```diff
@@ -0,0 +1,213 @@
+            return null;
+        }
+
+        traitInstancePtr = newDataSink(mWDMClientPtr, resourceIdentifier, profileId, instanceId, path);
+
+        if (traitInstancePtr == 0)
+        {
+            mCompHandler.onError(new Exception("Fail to create traitInstancePtr."));
+            Log.e(TAG, "traitInstancePtr is not ready");
+            return null;
+        }
+
+        isExist = mTraitMap.containsKey(traitInstancePtr);
+        if (isExist == false)
+        {
+            traitInstance = new GenericTraitUpdatableDataSinkImpl(traitInstancePtr, this);
+            mTraitMap.put(traitInstancePtr, traitInstance);
+        }
+        else
+        {
+            Log.d(TAG, "trait exists" + profileId);
+        }
+
+        Log.d(TAG, "there exists %d traits" + mTraitMap.size());
+
+        return mTraitMap.get(traitInstancePtr);
+    }
+
+    protected void removeDataSinkRef(long traitInstancePtr)
+    {
+        boolean isExist = false;
+        isExist = mTraitMap.containsKey(traitInstancePtr);
+        if (isExist == true)
+        {
+            mTraitMap.put(traitInstancePtr, null);
+        }
+    }
+
+    @Override
+    public void beginFlushUpdate()
+    {
+        if (mWDMClientPtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mWDMClientPtr is 0");
+            return;
+        }
+        beginFlushUpdate(mWDMClientPtr);
+    }
+
+    @Override
+    public void beginRefreshData()
+    {
+        if (mWDMClientPtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mWDMClientPtr is 0");
+            return;
+        }
+        beginRefreshData(mWDMClientPtr);
+    }
+
+    @Override
+    public CompletionHandler getCompletionHandler()
+    {
+        return mCompHandler;
+    }
+
+    @Override
+    public void setCompletionHandler(CompletionHandler compHandler)
+    {
+        mCompHandler = compHandler;
+    }
+
+    private void onError(Throwable err)
+    {
+        if (mCompHandler == null)
+        {
+            Log.e(TAG, "unexpected err, mCompHandler is null");
+            return;
+        }
+        mCompHandler.onError(err);
+    }
+
+    private void onFlushUpdateComplete()
+    {
+        if (mCompHandler == null)
+        {
+            Log.e(TAG, "unexpected err, mCompHandler is null");
+            return;
+        }
+        mCompHandler.onFlushUpdateComplete();
+    }
+
+    private void onRefreshDataComplete()
+    {
+        if (mCompHandler == null)
+        {
+            Log.e(TAG, "unexpected err, mCompHandler is null");
+            return;
+        }
+        mCompHandler.onRefreshDataComplete();
+    }
+
+    // ----- Protected Members -----
+
+    protected CompletionHandler mCompHandler;
+
+    protected void finalize() throws Throwable
+    {
+        super.finalize();
+        close();
+    }
+
+    // ----- Private Members -----
+
+    private long mWDMClientPtr;
+    private HashMap<Long, GenericTraitUpdatableDataSinkImpl> mTraitMap;
```

Comment:
nit: can be typed as `Map` instead of `HashMap` (interface vs implementation)

Reply:
Done, thanks